libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +243,166 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
+				constructorParams[dynamic_cast<ContractDefinition const*>(baseConstructor->scope())] =
+					listParamNames(baseConstructor, modifier->arguments());
+		}
 
-	for (ContractDefinition const* contract: reverse(_contract.annotation().linearizedBaseContracts))
+	for (ASTPointer<InheritanceSpecifier> const& base: _contract.baseContracts())
 	{
-		out <<
-			"\n// Begin state variable initialization for contract \"" <<
-			contract->name() <<
-			"\" (" <<
-			contract->stateVariables().size() <<
-			" variables)\n";
-
-		IRGeneratorForStatements generator{m_context, m_utils};
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+			base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			constructorParams[dynamic_cast<ContractDefinition const*>(baseConstructor->scope())] =
+				listParamNames(baseConstructor, base->arguments());
 	}
 
-	if (constructor)
+	return {generator.code(), constructorParams};
+}
+
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
+
+	return generator.code();
+}
+
+string IRGenerator::generateConstructorFunctions(ContractDefinition const& _contract)
+{
+	auto listAllParams = [&](
+		map<ContractDefinition const*, string> const& baseParams,
+		ContractDefinition const* nextContract = nullptr) -> string
 	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+		string paramsString;
+		if (nextContract != nullptr && baseParams.find(nextContract) != baseParams.end())
+			paramsString = baseParams.at(nextContract);
+		for (auto const& param: baseParams)
+			if (param.first != nextContract)
+				paramsString += (paramsString.empty() ? "" : ", ") + param.second;
+		return paramsString;
+	};
+
+	string constructorsCode;
+	map<ContractDefinition const*, string> baseConstructorParams;
+	for (size_t i = 0; i < _contract.annotation().linearizedBaseContracts.size(); ++i)
+	{
+		ContractDefinition const* contract = _contract.annotation().linearizedBaseContracts[i];
+		baseConstructorParams.erase(contract);
+
+		constructorsCode += m_context.functionCollector().createFunction(constructorName(*contract), [&]() {
+			Whiskers construtorFunction(R"(
```

Comment:
Maybe just cal it `template` - otherwise it might be confusing to see function calls to something called `constructorFunction`.

Reply:
I renamed to `t` as in other parts of code